flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -416,21 +416,31 @@ protected static <KEY, Context extends JobAutoScalerContext<KEY>> int scale(
                         // Optimize the case where newParallelism <= maxParallelism / 2
                         newParallelism > numKeyGroupsOrPartitions / 2
                                 ? numKeyGroupsOrPartitions
-                                : numKeyGroupsOrPartitions / 2,
+                                : numKeyGroupsOrPartitions / 2 + numKeyGroupsOrPartitions % 2,
                         upperBound);
 
+        boolean scalingRadical =
+                context.getConfiguration().get(AutoScalerOptions.SCALING_RADICAL_ENABLED);
+
         // When the shuffle type of vertex inputs contains keyBy or vertex is a source,
         // we try to adjust the parallelism such that it divides
         // the numKeyGroupsOrPartitions without a remainder => data is evenly spread across subtasks
         for (int p = newParallelism; p <= upperBoundForAlignment; p++) {
-            if (numKeyGroupsOrPartitions % p == 0) {
+            if (numKeyGroupsOrPartitions % p == 0
+                    ||
+                    // When scaling radical is enabled, Try to find the smallest parallelism that
+                    // can satisfy the
+                    // current consumption rate.
+                    (scalingRadical
+                            && numKeyGroupsOrPartitions / p
+                                    < numKeyGroupsOrPartitions / newParallelism)) {
```

Comment:
I think extracting this as a method `canMaximiseUtilisation` the intent is to make the intent of the condition easier to understand when working through the code.